# Model - Toby Liang

## I. Import Essential Libraries

In [1]:
# Arrays and dataframes
import numpy as np
import pandas as pd

# Visualization libraries
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

## II. Loading the Preprocessed Data

In [2]:
train_dataset = pd.read_csv("../data/preprocessed/Unsampled_train_dataset.csv")
train_dataset_undersampled = pd.read_csv("../data/preprocessed/Undersampled_train_dataset.csv")
train_dataset_oversampled = pd.read_csv("../data/preprocessed/Oversampled_train_dataset.csv")
train_dataset_smote = pd.read_csv("../data/preprocessed/SMOTE_train_dataset.csv")

train_datasets_dict = {"Unsampled": train_dataset, "Undersampled": train_dataset_undersampled, "Oversampled": train_dataset_oversampled, "SMOTE": train_dataset_smote}

## III. Implementing Deep Learning Model

In [24]:
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation
from keras.optimizers import Adam

In [25]:
# Model definition
def train_model(train_features, train_labels, val_features, val_labels, lr = 0.05, dp = 0.1, units = 64, batch_size = 32, epochs = 20):
    model = Sequential()

    # Input layer and first hidden
    model.add(Dense(units, activation = "relu", input_dim = train_features.shape[1]))
    model.add(Dropout(dp))
    
    # Additional hidden layer
    model.add(Dense(units, activation = "relu"))
    model.add(Dropout(dp))
    
    # Output layer
    model.add(Dense(1, activation = "sigmoid"))
    model.compile(optimizer = Adam(lr = lr), loss="binary_crossentropy", metrics = ["acc"])
    
    return (model, model.fit(train_features, train_labels, epochs = epochs, batch_size = batch_size, validation_data = (val_features, val_labels)))

### Model Hyperparameter Tuning

In [43]:
tuning_path = "./tuning/tuning_results.csv"
tuning_results = pd.read_csv(tuning_path)

In [44]:
tuning_results

Dataset  Learning Rate  Dropout  Units  Batch Size    F1 Avg  \
0      Unsampled         0.0100      0.1     64          32  0.516667   
1   Undersampled         0.0100      0.1     64          32  0.078009   
2    Oversampled         0.0100      0.1     64          32  0.614738   
3          SMOTE         0.0100      0.1     64          32  0.656100   
4      Unsampled         0.0100      0.2     64          32  0.342857   
5   Undersampled         0.0100      0.2     64          32  0.064499   
6    Oversampled         0.0100      0.2     64          32  0.645587   
7          SMOTE         0.0100      0.2     64          32  0.710582   
8      Unsampled         0.0100      0.3     64          32  0.338095   
9   Undersampled         0.0100      0.3     64          32  0.064615   
10   Oversampled         0.0100      0.3     64          32  0.653978   
11         SMOTE         0.0100      0.3     64          32  0.696742   
12     Unsampled         0.0100      0.1     32          32  0.680952   
13  Undersampled         0.0100      0.1     32          32  0.059307   
14   Oversampled         0.0100      0.1     32          32  0.701360   
15         SMOTE         0.0100      0.1     32          32  0.742228   
16     Unsampled         0.0050      0.2     32          32  0.855952   
17  Undersampled         0.0050      0.2     32          32  0.072299   
18   Oversampled         0.0050      0.2     32          32  0.737182   
19         SMOTE         0.0050      0.2     32          32  0.761806   
20     Unsampled         0.0100      0.2     32          32  0.493224   
21  Undersampled         0.0100      0.2     32          32  0.077722   
22   Oversampled         0.0100      0.2     32          32  0.632591   
23         SMOTE         0.0100      0.2     32          32  0.725293   
24     Unsampled         0.0100      0.0     32          32  0.822265   
25  Undersampled         0.0100      0.0     32          32  0.055810   
26   Oversampled         0.0100      0.0     32          32  0.761677   
27         SMOTE         0.0100      0.0     32          32  0.767719   
28     Unsampled         0.0050      0.0     32          32  0.839062   
29  Undersampled         0.0050      0.0     32          32  0.063186   
30   Oversampled         0.0050      0.0     32          32  0.835901   
31         SMOTE         0.0050      0.0     32          32  0.764508   
32     Unsampled         0.0050      0.1     32          32  0.858341   
33  Undersampled         0.0050      0.1     32          32  0.071521   
34   Oversampled         0.0050      0.1     32          32  0.756850   
35         SMOTE         0.0050      0.1     32          32  0.794005   
36     Unsampled         0.0010      0.0     32          32  0.853867   
37  Undersampled         0.0010      0.0     32          32  0.104473   
38   Oversampled         0.0010      0.0     32          32  0.855007   
39         SMOTE         0.0010      0.0     32          32  0.825400   
40     Unsampled         0.0010      0.1     32          32  0.867076   
41  Undersampled         0.0010      0.1     32          32  0.099392   
42   Oversampled         0.0010      0.1     32          32  0.789848   
43         SMOTE         0.0010      0.1     32          32  0.825312   
44     Unsampled         0.0010      0.2     32          32  0.841121   
45  Undersampled         0.0010      0.2     32          32  0.104051   
46   Oversampled         0.0010      0.2     32          32  0.796174   
47         SMOTE         0.0010      0.2     32          32  0.795802   
48     Unsampled         0.0005      0.2     32          32  0.829757   
49  Undersampled         0.0005      0.2     32          32  0.135940   
50   Oversampled         0.0005      0.2     32          32  0.788711   
51         SMOTE         0.0005      0.2     32          32  0.798830   
52     Unsampled         0.0005      0.1     32          32  0.848659   
53  Undersampled         0.0005      0.1     32          32  0.1182

In [28]:
from sklearn.metrics import f1_score, recall_score, precision_score, accuracy_score

def compute_stats(stat_lst):
    return (np.average(stat_lst), np.std(stat_lst))

def save_plot(hist, key, lr, dp, units, batch_size):
    title = "{}-lr{}-dp{}-units{}-batchsize{}".format(key, lr, dp, units, batch_size)
    fig = plt.figure(figsize = (15, 5))
    plt.title(title)
    plt.plot(hist.history["loss"], label = "Training loss")
    plt.plot(hist.history["val_loss"], label = "Validation loss")
    plt.xlabel("Epochs")
    plt.ylabel("Loss")
    plt.legend()
    plt.savefig("./tuning/plots/{}.png".format(title))
    plt.close(fig)

def tune_hyperparameters(key, lr, dp, units, batch_size, iterations):
        prec_lst, rec_lst, f1_lst, acc_lst = [], [], [], []

        for i in range(iterations):
            print("Dataset: {}, Iteration: {}".format(key, i + 1))

            model, hist = train_model(train_features_dict[key].values, train_labels_dict[key].values, val_features.values, val_labels.values, lr = lr, dp = dp, units = units, batch_size = batch_size)
            predictions = model.predict_classes(test_features)
            
            # Plot loss
            save_plot(hist, key, lr, dp, units, batch_size)

            # Append evaluation scores to lists
            f1_lst.append(f1_score(test_labels, predictions))
            rec_lst.append(recall_score(test_labels, predictions))
            prec_lst.append(precision_score(test_labels, predictions))
            acc_lst.append(accuracy_score(test_labels, predictions))

        # Compute stats
        f1_avg, f1_std = compute_stats(f1_lst)
        rec_avg, rec_std = compute_stats(rec_lst)
        prec_avg, prec_std = compute_stats(prec_lst)
        acc_avg, acc_std = compute_stats(acc_lst)

        return pd.DataFrame({"Dataset": [key], "Learning Rate": [lr], "Dropout": [dp], "Units": [units], "Batch Size":[batch_size], "F1 Avg": [f1_avg], "F1 Std": [f1_std], "Recall Avg": [rec_avg], "Recall Std": [rec_std], "Precision Avg": [prec_avg], "Precision Std": [prec_std], "Accuracy Avg": [acc_avg], "Accuracy Std": [acc_std], "Iterations": [iterations]})

In [39]:
lr = 0.001
dp = 0.1
units = 32
batch_size = 16
iterations = 5

for key in train_datasets_dict:
    result = tune_hyperparameters(key, lr, dp, units, batch_size, iterations)
    tuning_results = tuning_results.append(result)

Dataset: Unsampled, Iteration: 1
Train on 261452 samples, validate on 8087 samples
Epoch 1/20
261452/261452 [==============================] - 28s 107us/step - loss: 0.0072 - acc: 0.9985 - val_loss: 0.0279 - val_acc: 0.9983
Epoch 2/20
261452/261452 [==============================] - 23s 89us/step - loss: 0.0034 - acc: 0.9993 - val_loss: 0.0279 - val_acc: 0.9983
Epoch 3/20
261452/261452 [==============================] - 23s 89us/step - loss: 0.0031 - acc: 0.9994 - val_loss: 0.0279 - val_acc: 0.9983
Epoch 4/20
261452/261452 [==============================] - 24s 90us/step - loss: 0.0031 - acc: 0.9994 - val_loss: 0.0279 - val_acc: 0.9983
Epoch 5/20
261452/261452 [==============================] - 23s 89us/step - loss: 0.0028 - acc: 0.9994 - val_loss: 0.0279 - val_acc: 0.9983
Epoch 6/20
261452/261452 [==============================] - 23s 88us/step - loss: 0.0029 - acc: 0.9994 - val_loss: 0.0279 - val_acc: 0.9983
Epoch 7/20
261452/261452 [==============================] - 24s 90us/step - 

868/868 [==============================] - 0s 525us/step - loss: 0.1202 - acc: 0.9528 - val_loss: 0.0319 - val_acc: 0.9980
Epoch 15/20
868/868 [==============================] - 0s 519us/step - loss: 0.1084 - acc: 0.9562 - val_loss: 0.0319 - val_acc: 0.9980
Epoch 16/20
868/868 [==============================] - 0s 529us/step - loss: 0.1128 - acc: 0.9528 - val_loss: 0.0339 - val_acc: 0.9978
Epoch 17/20
868/868 [==============================] - 0s 527us/step - loss: 0.0989 - acc: 0.9574 - val_loss: 0.0319 - val_acc: 0.9980
Epoch 18/20
868/868 [==============================] - 0s 529us/step - loss: 0.1053 - acc: 0.9574 - val_loss: 0.0350 - val_acc: 0.9978
Epoch 19/20
868/868 [==============================] - 0s 525us/step - loss: 0.0997 - acc: 0.9597 - val_loss: 0.0319 - val_acc: 0.9979
Epoch 20/20
868/868 [==============================] - 0s 528us/step - loss: 0.0939 - acc: 0.9597 - val_loss: 0.0349 - val_acc: 0.9978
Dataset: Undersampled, Iteration: 2
Train on 868 samples, validate 

Epoch 11/20
522036/522036 [==============================] - 51s 98us/step - loss: 0.0063 - acc: 0.9986 - val_loss: 0.0279 - val_acc: 0.9983
Epoch 12/20
522036/522036 [==============================] - 52s 99us/step - loss: 0.0065 - acc: 0.9986 - val_loss: 0.0279 - val_acc: 0.9983
Epoch 13/20
522036/522036 [==============================] - 51s 97us/step - loss: 0.0062 - acc: 0.9987 - val_loss: 0.0279 - val_acc: 0.9983
Epoch 14/20
522036/522036 [==============================] - 51s 98us/step - loss: 0.0060 - acc: 0.9986 - val_loss: 0.0279 - val_acc: 0.9983
Epoch 15/20
522036/522036 [==============================] - 51s 98us/step - loss: 0.0061 - acc: 0.9987 - val_loss: 0.0279 - val_acc: 0.9983
Epoch 16/20
522036/522036 [==============================] - 52s 99us/step - loss: 0.0059 - acc: 0.9987 - val_loss: 0.0288 - val_acc: 0.9980
Epoch 17/20
522036/522036 [==============================] - 51s 98us/step - loss: 0.0060 - acc: 0.9987 - val_loss: 0.0283 - val_acc: 0.9980
Epoch 18/20
5

522036/522036 [==============================] - 62s 118us/step - loss: 0.0073 - acc: 0.9983 - val_loss: 0.0279 - val_acc: 0.9983
Epoch 4/20
522036/522036 [==============================] - 60s 115us/step - loss: 0.0066 - acc: 0.9986 - val_loss: 0.0279 - val_acc: 0.9983
Epoch 5/20
522036/522036 [==============================] - 61s 116us/step - loss: 0.0061 - acc: 0.9987 - val_loss: 0.0279 - val_acc: 0.9983
Epoch 6/20
522036/522036 [==============================] - 61s 116us/step - loss: 0.0057 - acc: 0.9988 - val_loss: 0.0279 - val_acc: 0.9983
Epoch 7/20
522036/522036 [==============================] - 61s 116us/step - loss: 0.0049 - acc: 0.9990 - val_loss: 0.0279 - val_acc: 0.9983
Epoch 8/20
522036/522036 [==============================] - 60s 116us/step - loss: 0.0048 - acc: 0.9991 - val_loss: 0.0279 - val_acc: 0.9983
Epoch 9/20
522036/522036 [==============================] - 64s 122us/step - loss: 0.0046 - acc: 0.9991 - val_loss: 0.0279 - val_acc: 0.9983
Epoch 10/20
522036/52203

Epoch 20/20
522036/522036 [==============================] - 69s 132us/step - loss: 0.0040 - acc: 0.9992 - val_loss: 0.0279 - val_acc: 0.9983


In [40]:
tuning_results

Dataset  Learning Rate  Dropout  Units  Batch Size    F1 Avg  \
0      Unsampled         0.0100      0.1     64          32  0.516667   
1   Undersampled         0.0100      0.1     64          32  0.078009   
2    Oversampled         0.0100      0.1     64          32  0.614738   
3          SMOTE         0.0100      0.1     64          32  0.656100   
4      Unsampled         0.0100      0.2     64          32  0.342857   
5   Undersampled         0.0100      0.2     64          32  0.064499   
6    Oversampled         0.0100      0.2     64          32  0.645587   
7          SMOTE         0.0100      0.2     64          32  0.710582   
8      Unsampled         0.0100      0.3     64          32  0.338095   
9   Undersampled         0.0100      0.3     64          32  0.064615   
10   Oversampled         0.0100      0.3     64          32  0.653978   
11         SMOTE         0.0100      0.3     64          32  0.696742   
12     Unsampled         0.0100      0.1     32          32  0.680952   
13  Undersampled         0.0100      0.1     32          32  0.059307   
14   Oversampled         0.0100      0.1     32          32  0.701360   
15         SMOTE         0.0100      0.1     32          32  0.742228   
16     Unsampled         0.0050      0.2     32          32  0.855952   
17  Undersampled         0.0050      0.2     32          32  0.072299   
18   Oversampled         0.0050      0.2     32          32  0.737182   
19         SMOTE         0.0050      0.2     32          32  0.761806   
20     Unsampled         0.0100      0.2     32          32  0.493224   
21  Undersampled         0.0100      0.2     32          32  0.077722   
22   Oversampled         0.0100      0.2     32          32  0.632591   
23         SMOTE         0.0100      0.2     32          32  0.725293   
24     Unsampled         0.0100      0.0     32          32  0.822265   
25  Undersampled         0.0100      0.0     32          32  0.055810   
26   Oversampled         0.0100      0.0     32          32  0.761677   
27         SMOTE         0.0100      0.0     32          32  0.767719   
28     Unsampled         0.0050      0.0     32          32  0.839062   
29  Undersampled         0.0050      0.0     32          32  0.063186   
30   Oversampled         0.0050      0.0     32          32  0.835901   
31         SMOTE         0.0050      0.0     32          32  0.764508   
32     Unsampled         0.0050      0.1     32          32  0.858341   
33  Undersampled         0.0050      0.1     32          32  0.071521   
34   Oversampled         0.0050      0.1     32          32  0.756850   
35         SMOTE         0.0050      0.1     32          32  0.794005   
36     Unsampled         0.0010      0.0     32          32  0.853867   
37  Undersampled         0.0010      0.0     32          32  0.104473   
38   Oversampled         0.0010      0.0     32          32  0.855007   
39         SMOTE         0.0010      0.0     32          32  0.825400   
40     Unsampled         0.0010      0.1     32          32  0.867076   
41  Undersampled         0.0010      0.1     32          32  0.099392   
42   Oversampled         0.0010      0.1     32          32  0.789848   
43         SMOTE         0.0010      0.1     32          32  0.825312   
44     Unsampled         0.0010      0.2     32          32  0.841121   
45  Undersampled         0.0010      0.2     32          32  0.104051   
46   Oversampled         0.0010      0.2     32          32  0.796174   
47         SMOTE         0.0010      0.2     32          32  0.795802   
48     Unsampled         0.0005      0.2     32          32  0.829757   
49  Undersampled         0.0005      0.2     32          32  0.135940   
50   Oversampled         0.0005      0.2     32          32  0.788711   
51         SMOTE         0.0005      0.2     32          32  0.798830   
0      Unsampled         0.0005      0.1     32          32  0.848659   
0   Undersampled         0.0005      0.1     32          32  0.1182

In [41]:
tuning_results.to_csv(tuning_path, index = False, columns = ["Dataset", "Learning Rate", "Dropout", "Units", "Batch Size", "F1 Avg", "F1 Std", "Recall Avg", "Recall Std", "Precision Avg", "Precision Std", "Accuracy Avg", "Accuracy Std", "Iterations"])

### Plotting Model Loss

In [ ]:
# Fitting models
model_dict = {}
hist_dict = {}
for key in train_datasets_dict:
    model, hist = train_model(train_features_dict[key].values, train_labels_dict[key].values, val_features.values, val_labels.values)
    model_dict[key] = model
    hist_dict[key] = hist

In [ ]:
# Plotting models
for key, val in hist_dict.items():
    plt.figure(figsize = (15, 5))
    plt.title(key)
    plt.plot(hist_dict[key].history["loss"], label = "Training loss")
    plt.plot(hist_dict[key].history["val_loss"], label = "Validation loss")
    plt.xlabel("Epochs")
    plt.ylabel("Loss")
    plt.legend()
    plt.show()

### Evaluating Models

In [ ]:
print("Test Dataset Fraud Count\n{}".format(test_labels["label"].value_counts()))

In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix
for key, val in hist_dict.items():
    predictions = model_dict[key].predict_classes(test_features)
    acc = accuracy_score(test_labels, predictions)
    prec = precision_score(test_labels, predictions)
    rec = recall_score(test_labels, predictions)
    f1 = f1_score(test_labels, predictions)
    c_matrix = confusion_matrix(test_labels, predictions)
    
    print("{} Metrics Report".format(key))
    metrics_dict = {"Accuracy": acc, "Precision": prec, "Recall": rec, "F1": f1}
    for name, metric in metrics_dict.items():
        print("{}: {}".format(name, round(metric, 4)))
    print("True Negative: {}".format(c_matrix.ravel()[0]))
    print("False Positive: {}".format(c_matrix.ravel()[1]))
    print("False Negative: {}".format(c_matrix.ravel()[2]))
    print("True Positive: {}".format(c_matrix.ravel()[3]))
    print(c_matrix)
    print()

## Results

**lr = 0.05, epochs = 50
Test Dataset Fraud Count
0    14162
1       25
Name: label, dtype: int64**

Unsampled Metrics Report
- Accuracy: 0.9982
Precision: 0.0
Recall: 0.0
F1: 0.0
- True Negative: 14162
False Positive: 0
False Negative: 25
True Positive: 0
- [[14162     0]
 [   25     0]]

Undersampled Metrics Report
- Accuracy: 0.9394
Precision: 0.025
Recall: 0.88
F1: 0.0487
- True Negative: 13305
False Positive: 857
False Negative: 3
True Positive: 22
- [[13305   857]
 [    3    22]]

Oversampled Metrics Report
- Accuracy: 0.9991
Precision: 0.75
Recall: 0.72
F1: 0.7347
- True Negative: 14156
False Positive: 6
False Negative: 7
True Positive: 18
- [[14156     6]
 [    7    18]]

SMOTE Metrics Report
- Accuracy: 0.9807
Precision: 0.0751
Recall: 0.88
F1: 0.1384
- True Negative: 13891
False Positive: 271
False Negative: 3
True Positive: 22
- [[13891   271]
 [    3    22]]
 
**lr = 0.05, epochs = 20
Test Dataset Fraud Count
0    14162
1       25
Name: label, dtype: int64**

Unsampled Metrics Report
- Accuracy: 0.9982
Precision: 0.0
Recall: 0.0
F1: 0.0
- True Negative: 14162
False Positive: 0
False Negative: 25
True Positive: 0
- [[14162     0]
 [   25     0]]

Undersampled Metrics Report
- Accuracy: 0.9385
Precision: 0.0236
Recall: 0.84
F1: 0.0459
- True Negative: 13293
False Positive: 869
False Negative: 4
True Positive: 21
- [[13293   869]
 [    4    21]]

Oversampled Metrics Report
- Accuracy: 0.9993
Precision: 0.8261
Recall: 0.76
F1: 0.7917
- True Negative: 14158
False Positive: 4
False Negative: 6
True Positive: 19
- [[14158     4]
 [    6    19]]

SMOTE Metrics Report
- Accuracy: 0.9867
Precision: 0.1063
Recall: 0.88
F1: 0.1897
- True Negative: 13977
False Positive: 185
False Negative: 3
True Positive: 22
- [[13977   185]
 [    3    22]]

**lr = 0.05, epochs = 10
Test Dataset Fraud Count
0    14162
1       25
Name: label, dtype: int64**

Unsampled Metrics Report
Accuracy: 0.9982
Precision: 0.0
Recall: 0.0
F1: 0.0
True Negative: 14162
False Positive: 0
False Negative: 25
True Positive: 0
[[14162     0]
 [   25     0]]

Undersampled Metrics Report
- Accuracy: 0.9438
Precision: 0.027
Recall: 0.88
F1: 0.0523
- True Negative: 13368
False Positive: 794
False Negative: 3
True Positive: 22
- [[13368   794]
 [    3    22]]

Oversampled Metrics Report
- Accuracy: 0.9982
Precision: 0.0
Recall: 0.0
F1: 0.0
- True Negative: 14162
False Positive: 0
False Negative: 25
True Positive: 0
- [[14162     0]
 [   25     0]]

SMOTE Metrics Report
- Accuracy: 0.9937
Precision: 0.1923
Recall: 0.8
F1: 0.3101
- True Negative: 14078
False Positive: 84
False Negative: 5
True Positive: 20
- [[14078    84]
 [    5    20]]

- Precision = Actual / Predicted ( Wrong Predictions )
    - TP / (TP + FP)
- Recall = Predicted / Actual ( Missed Predictions )
    - TP / (TP + FN)